x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java

```diff
@@ -0,0 +1,247 @@
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.ml.queries.WeightedTokensThreshold.TOKENS_THRESHOLD_FIELD;
+
+public class WeightedTokensQueryBuilder extends AbstractQueryBuilder<WeightedTokensQueryBuilder> {
+    public static final String NAME = "weighted_tokens";
+
+    public static final ParseField TOKENS_FIELD = new ParseField("tokens");
+    private final String fieldName;
+    private final Set<WeightedToken> tokens;
+    private final WeightedTokensThreshold threshold;
+
+    public WeightedTokensQueryBuilder(String fieldName, Set<WeightedToken> tokens) {
+        this(fieldName, tokens, null);
+    }
+
+    public WeightedTokensQueryBuilder(String fieldName, Set<WeightedToken> tokens, @Nullable WeightedTokensThreshold threshold) {
+        this.fieldName = Objects.requireNonNull(fieldName, "[" + NAME + "] requires a fieldName");
+        this.tokens = Objects.requireNonNull(tokens, "[" + NAME + "] requires tokens");
+        if (tokens.isEmpty()) {
+            throw new IllegalArgumentException("[" + NAME + "] requires at least one token");
+        }
+        this.threshold = Objects.requireNonNullElse(threshold, new WeightedTokensThreshold());
+    }
+
+    public WeightedTokensQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.tokens = in.readCollectionAsSet(WeightedToken::new);
+        this.threshold = in.readOptionalWriteable(WeightedTokensThreshold::new);
+    }
+
+    public String getFieldName() {
+        return fieldName;
+    }
+
+    @Nullable
+    public WeightedTokensThreshold getThreshold() {
+        return threshold;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeCollection(tokens);
+        out.writeOptionalWriteable(threshold);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.startObject(TOKENS_FIELD.getPreferredName());
+        for (var token : tokens) {
+            token.toXContent(builder, params);
+        }
+        builder.endObject();
+        if (threshold != null) {
+            threshold.toXContent(builder, params);
+        }
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    private float getAverageTokenFreqRatio(IndexReader reader, int fieldDocCount) throws IOException {
+        int numUniqueTokens = 0;
+        for (var leaf : reader.getContext().leaves()) {
+            var terms = leaf.reader().terms(fieldName);
+            if (terms != null) {
+                numUniqueTokens = (int) Math.max(terms.size(), numUniqueTokens);
```

Comment:
The alternative would be to compute an average token frequency per shard and then compute the weighted (by doc count) average of these averages.

Reply:
> Is this the maximum number of unique tokens for any shard of the data?

Yes, we use it as a proxy since we don't have the unique count inter-segment.

Reply:
I thought some more about this. Whether using average per shard frequency ratio or max frequency ratio and total document count works depends on what you assume. In particular, if the token set overlaps are large (which is actualy likely) this approach is better.